examples/lock-app/silabs/src/LockManager.cpp

```diff
@@ -353,34 +333,78 @@ bool LockManager::Unbolt(chip::EndpointId endpointId, const Nullable<chip::Fabri
 
     userIndex--;
 
     VerifyOrReturnValue(IsValidUserIndex(userIndex), false);
 
+    VerifyOrReturnValue(kInvalidEndpointId != endpointId, false);
+
     ChipLogProgress(Zcl, "Door Lock App: LockManager::GetUser [endpoint=%d,userIndex=%hu]", endpointId, userIndex);
 
-    const auto & userInDb = mLockUsers[userIndex];
+    chip::StorageKeyName userKey = LockUserEndpoint(userIndex, endpointId);
+
+    uint16_t size = LockUserInfoSize;
 
-    user.userStatus = userInDb.userStatus;
-    if (UserStatusEnum::kAvailable == user.userStatus)
+    error = mStorage->SyncGetKeyValue(userKey.KeyName(), &mUserInStorage, size);
+    // If no data is found at user key
+    if (error == CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND)
     {
-        ChipLogDetail(Zcl, "Found unoccupied user [endpoint=%d]", endpointId);
+        user.userStatus = UserStatusEnum::kAvailable;
+
+        ChipLogDetail(Zcl, "No user data found");
         return true;
     }
+    // Else if KVS read was successful
+    else if (error == CHIP_NO_ERROR)
+    {
+        user.userStatus = mUserInStorage.userStatus;
 
-    user.userName       = chip::CharSpan(userInDb.userName.data(), userInDb.userName.size());
-    user.credentials    = chip::Span<const CredentialStruct>(mCredentials[userIndex], userInDb.credentials.size());
-    user.userUniqueId   = userInDb.userUniqueId;
-    user.userType       = userInDb.userType;
-    user.credentialRule = userInDb.credentialRule;
+        if (mUserInStorage.userStatus == UserStatusEnum::kAvailable)
+        {
+            ChipLogDetail(Zcl, "Found unoccupied user [endpoint=%d]", endpointId);
+            return true;
+        }
+    }
+    else
+    {
+        ChipLogError(Zcl, "Error reading from KVS key");
+        return false;
+    }
+
+    VerifyOrReturnValue(mUserInStorage.currentCredentialCount <= kMaxCredentialsPerUser, false);
+
+    user.userName       = chip::CharSpan(mUserInStorage.userName, mUserInStorage.userNameSize);
+    user.userUniqueId   = mUserInStorage.userUniqueId;
+    user.userType       = mUserInStorage.userType;
+    user.credentialRule = mUserInStorage.credentialRule;
     // So far there's no way to actually create the credential outside Matter, so here we always set the creation/modification
     // source to Matter
     user.creationSource     = DlAssetSource::kMatterIM;
-    user.createdBy          = userInDb.createdBy;
+    user.createdBy          = mUserInStorage.createdBy;
     user.modificationSource = DlAssetSource::kMatterIM;
-    user.lastModifiedBy     = userInDb.lastModifiedBy;
+    user.lastModifiedBy     = mUserInStorage.lastModifiedBy;
+
+    // Get credential struct from nvm3
+    chip::StorageKeyName credentialKey = LockUserCredentialMap(userIndex);
+
+    uint16_t credentialSize = static_cast<uint16_t>(CredentialStructSize * mUserInStorage.currentCredentialCount);
+
+    error = mStorage->SyncGetKeyValue(credentialKey.KeyName(), mCredentials, credentialSize);
```

Comment:
If we get fewer than credentialSize bytes out (because fewer are stored) call will succeed, but we will have uninitialized bits that we include in our span, right?

Reply:
Added checks after each SyncGetKeyValue to check the size outparam